Feature sockets (#131)
samply/beam: README.md

Author: Could for the live of me not figure out a way to do it with curl or netcat. But I think python is nicer to read anyway

}
```

### Socket connections
> Note: Only avalibale on builds with the feature `sockets` enabled. Both proxy and broker need to be build with this flag. There are also prebuild docker images avalible with this feature.

All api request require the normal authenticated header see the [getting started section](#getting-started).

To create initialize a socket connection to an app with a given beam id i.e `app2.proxy2.broker`:

Method: `POST`  
URL: `/v1/sockets/<BEAM_ID>`  

> Header: `Upgrade` is required

Author: Yeah so I am not sure if the hyper client sets `Connection: upgrade` automatically if `Upgrade` is set. But the value of upgrade is not used we don't care about what is sent through here. But hyper insists on having it make an upgrade, maybe I could drop this requirement and somehow manually set the header before hyper reads it but that might not be possible. Will look into it.

> Header: `Upgrade` is required

Reviewer: Yes, I would provide e.g. "Upgrade: ValueNotUsed", but please add "Connection: upgrade" to be standard compleant (even if we don't use the header). What do you think @lablans?

> Header: `Upgrade` is required

Reviewer 2: Agreed, also please ask "upstream" in the hyper repository and link to the issue, so everyone can understand why we opted to do it this way.

After this you will be automatically connected to the other app if they answer your request.

> To recieve connections from other apps the server needs to poll for incoming connections.

Reviewer: ```suggestion
#### Receive and answer a socket request
To receive socket connections, the Beam.Proxy needs to be polled for incoming connections.
```

This endpoint also supports the [long polling](#long-polling-api-access) query string semantics.

> Method: `GET`  
> URL: `/v1/sockets`

Reviewer: ```suggestion
Method: `GET`  
URL: `/v1/sockets`
Parameters:
 * The same parameters as for long-polling, i.e. to, from, filter=todo, wait_count, and wait_time are supported.
```

> This will return a json object like the following

Reviewer: ```suggestion
Returns an array of JSON objects:
```

> ```

Reviewer: ```suggestion
``` json
```

[
    {
        "from": "app1.proxy1.broker",
        "to": ["app2.proxy2.broker"]

>         "id": "Some uuid v4",

Reviewer: ```suggestion
        "id": "<socket_uuid>",
```

        "ttl": "60s",
    }
]
```

> The client can then proceed to connect to the socket with:

Reviewer: ```suggestion
#### Connecting to a socket request
After the connection negotion above, the App can proceed to connect to the socket:
```

Method: GET  

> URL: `/v1/sockets/<id>`

Reviewer: ```suggestion
URL: `/v1/sockets/<socket_uuid>`
```
